Add a protection for possible mismatch in sensor IDs

paolafer opened the pull request:
This PR adds a protection that raises an error if the nexus file has PMT IDs which are not in the MySQL database.
Closes #779.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -451,6 +453,16 @@ def mcsensors_from_file(paths     : List[str],
                                                        db_file    = db_file   ,
                                                        run_no     = run_number)
+        sns_pos_tbl = 'MC/sns_positions'
+        if is_oldformat_file(file_name): sns_pos_tbl = 'MC/sensor_positions'
```
```suggestion
        sns_pos_tbl = 'MC/sensor_positions' if is_oldformat_file(file_name) else 'MC/sns_positions'
```
more idiomatic

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -451,6 +453,16 @@ def mcsensors_from_file(paths     : List[str],
+        if is_oldformat_file(file_name): sns_pos_tbl = 'MC/sensor_positions'
+
+        nexus_sns_pos       = pd.read_hdf(file_name, sns_pos_tbl)
+        nexus_pmt_ids       = nexus_sns_pos[nexus_sns_pos.sensor_id < 100].sensor_id.values
```
This is dangerous and I believe there is another instance in the code where we rely on a magic number. We should write the sensor type in this table in nexus. For now, we can add a comment that clarifies this number.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -451,6 +453,16 @@ def mcsensors_from_file(paths     : List[str],
+        nexus_pmt_ids       = nexus_sns_pos[nexus_sns_pos.sensor_id < 100].sensor_id.values
+        nexus_pmt_arg_in_db = np.argwhere([sid in pmt_ids for sid in nexus_pmt_ids])
+
+        if len(nexus_pmt_ids) != len(nexus_pmt_arg_in_db):
```
A more concise way to express that every number in a sequence is present in another sequence, using a `pd.Series`, is:
```
nexus_pmt_ids = nexus_sns_pos[nexus_sns_pos.sensor_id < 100].sensor_id
if nexus_pmt_ids.isin(pmt_ids).all():
```

gonzaponte reviewed (changes requested):
Some minor comments.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -453,6 +455,15 @@ def mcsensors_from_file(paths     : List[str],
+        sns_pos_tbl = 'MC/sensor_positions' if is_oldformat_file(file_name) else 'MC/sns_positions'
+
+        nexus_sns_pos       = pd.read_hdf(file_name, sns_pos_tbl)
+        nexus_pmt_ids       = nexus_sns_pos[nexus_sns_pos.sensor_id < 100].sensor_id ## sensors with IDs below 100 are PMTs, the rest are SiPMs
```
```suggestion
        nexus_sns_pos = pd.read_hdf(file_name, sns_pos_tbl)
        nexus_pmt_ids = nexus_sns_pos[nexus_sns_pos.sensor_id < 100].sensor_id ## sensors with IDs below 100 are PMTs, the rest are SiPMs
```
just cosmetics

gonzaponte reviewed (commented):
I would like to add a test for this.
Perhaps creating a file on the fly with wrong sensor ids and verify that it raises the expected exception. what do you think?

gonzaponte commented on invisible_cities/cities/components_test.py:
```diff
@@ -318,3 +319,14 @@ def test_check_max_time_units():
+
+    with raises(SensorIDMismatch):
+        sns_gen = mcsensors_from_file([file_in], 'next100', run_number, rate)
+        _ = next(sns_gen)
```
```suggestion
    sns_gen = mcsensors_from_file([file_in], 'next100', run_number, rate)
    with raises(SensorIDMismatch):
        next(sns_gen)
```
to make it more clear where the exception is raised.

gonzaponte commented on invisible_cities/cities/components_test.py:
```diff
@@ -318,3 +319,14 @@ def test_check_max_time_units():
 
+def test_read_wrong_pmt_ids(ICDATADIR):
+    """
+    Check if an error is raised when wrong PMT IDs are read from nexus file
```
Maybe explain that the nexus file contains one or more sensor IDs that are not present in the database, that we expect this to produce an error, and that therefore we check for that.

paolafer replied:
Done.

gonzaponte reviewed (commented):
I'm happy with this. Just two minor comments and I'll approve it.

gonzaponte reviewed (approved):
Fixes a long-standing issue. Great job!

paolafer committed:
Add a protection for possible mismatch in sensor IDs

Between nexus file and MySQL database

paolafer committed:
Make code more concise

paolafer committed:
Cosmetics

paolafer committed:
Add test and use PMT string instead of number to identify PMTs

I realized that, if there is an error in nexus, the PMT IDs couls be
larger than 100, so I decided to use the PMT insensitive-case string instead.

paolafer committed:
Do the check only for files in new format

paolafer committed:
Avoid going through numpy and keep using pandas instead

paolafer committed:
Make the test clearer.
